moved to the enricher model to avoid the need of tansmitting cats and…

kailuowang opened the pull request:
… shapeless data structures
fixes #67 
Per conversation with @joprice we need to address the following
- [x] replace `PathPattern` and `ActorPath` with in-house data representation to get rid of play and akka dependency. 
- [x] move protocol classes together under a package
- [x] use Array[Byte] instead of AnyContent, moving JSON parsing to the service side.

joprice commented on build.sbt:
```diff
@@ -70,7 +70,7 @@ val commonSettings = Seq(
     "-deprecation",
-    "-unchecked",
-    "-Xlint"
+//    "-Xlint",
```
Why disable linting?

kailuowang replied:
It complains of the auto tupling introduced by the magnet pattern, is there any other way to avoid that?

lvauthrin replied:
I think you can disable specific warnings although it might still be of a broader scope than you need.

joprice replied:
I think you can do `-Xlint` followed by the setting you want to disable with a single dash before it: "-Xlint:-adapted-args". I've also seen it combined like this: `-Xlint:_,-adapted-args` (https://github.com/search?utf8=%E2%9C%93&q=language%3Ascala++%22-Xlint%3A_%2C-adapted-args%22+&type=Code&ref=searchresults)

joprice commented on example/frontend/conf/application.conf:
```diff
@@ -66,8 +66,7 @@ akka.extensions = [ "akka.cluster.metrics.ClusterMetricsExtension" , "akka.clust
 asobu {
-  bridgePropsClass = util.KanaloaBridge
-  extractorDefinitionInterpreterClass = cross.MyInterpreter
+  endpointFactoryProviderClass = util.ExampleEndpointFactoryProvider
```
Do we really need to provide all this bootstrapping? It's starting to feel like a Spring app.

kailuowang replied:
This is required only when Play runtime DI is used.

joprice commented:
There's a lot going on in this PR. Can we break out some of the changes? There some library upgrades which seem independent enough. And removing all the binary sigar libs that were committed for some reason could be done by itself.

joprice commented on distributed/src/main/scala/asobu/distributed/gateway/Endpoint.scala:
```diff
@@ -70,35 +68,73 @@ case class Endpoint(
+  def handle(request: GateWayRequest): Future[Result] = {
     import akka.pattern.ask
-    import ExecutionContext.Implicits.global
+    //    import ExecutionContext.Implicits.global
```
Dead code

nsauro commented:
I agree with joseph, this PR is pretty large, can it be broken down a bit?

kailuowang committed:
moved to the enricher model

kailuowang commented:
@joprice broke out the build change to another PR #70, this PR is on top of that one.

kailuowang committed:
refactor to simplify code

kailuowang committed:
wip moving classes to protocol

kailuowang committed:
Take full control of EndpointDefinition data presentation

kailuowang committed:
moved DRequest and DResult into protocol

kailuowang committed:
transfer raw body, i.e. move json parsing to the service side.

kailuowang commented:
@joprice @nsauro @lvauthrin ready to be reviewed. I addressed the feedback I got so far. 
It's hard to divide this set of changes to different PRs and each pass builds. But I think we can review the changes per commit, which is smaller.

lvauthrin commented:
For a change this big, I think it would make sense to hold a short meeting so that you can give an overview of what you were trying to achieve so that there's more context for reviewers.

joprice commented on dsl/src/main/scala/asobu/dsl/Extractor.scala:
```diff
@@ -31,19 +31,29 @@ trait ExtractorOps {
+  //todo: use Magnet pattern for the creation shortcuts here
+  implicit def of[TFrom, T](f: TFrom ⇒ ExtractResult[T]): Extractor[TFrom, T] = Kleisli(f)
+
+  @deprecated("user `of` instead", "0.5.0")
```
typo: user -> use

joprice commented on distributed/src/main/scala/asobu/distributed/package.scala:
```diff
@@ -1,5 +1,15 @@
 package object distributed {
   type Headers = Seq[(String, String)] //Use Seq instead of Map to better ensure serialization
+
+  type Body = play.api.mvc.AnyContent //use Array[Byte] or other serializable data structure
```
Is this still being used since you introduced usage of `RawBuffer` / `Array[Byte]`?

kailuowang committed:
feedback

kailuowang commented:
@joprice any other feedback? @lvauthrin @nsauro?

nsauro commented on distributed/src/main/scala/asobu/distributed/protocol/DRequest.scala:
```diff
@@ -0,0 +1,65 @@
+import scala.concurrent.Future
+
+@SerialVersionUID(1L)
+case class HttpStatus(code: Int) extends AnyVal
```
We are pulling Play! into each service now, not sure if that was the case before.  However, since we are using Play! in both front and back end..does it make sense to just use their classes wholesale, and not provide wrappers?

nsauro replied:
By classes, i mean their http classes

kailuowang replied:
This way we can use different version of play at each end. Total control.

nsauro commented on distributed/src/main/scala/asobu/distributed/protocol/EndPointDefinition.scala:
```diff
@@ -0,0 +1,100 @@
+@SerialVersionUID(1L)
+case class Verb(value: String)
+
+sealed trait PathPart
```
Are these just copies of what is in play.routing?  Can we just use those, since we already have that dependency, or is there some issue with doing that?

kailuowang replied:
same as above.

nsauro commented on distributed/src/main/scala/asobu/distributed/service/DRequestExtractor.scala:
```diff
@@ -0,0 +1,154 @@
+import scala.language.implicitConversions
+import scala.util.{Failure, Success, Try}
+
+object DRequestExtractor extends ExtractorFunctions {
```
Just out of curiosity, do we have any idea on how much overhead this introduces?  This is pretty "hot" code, and executes on every single request.

kailuowang replied:
These are simple functions, the same as the ones we are using in rest-apis. And even better, they are distributed to the services now.

nsauro commented on distributed/src/test/scala/asobu/distributed/util/MockRoute.scala:
```diff
@@ -2,7 +2,7 @@ package asobu.distributed.util
 import play.routes.compiler._
 import play.api.http.HttpVerbs._
-
+//todo:
```
is this a commented out todo? :)

nsauro commented on distributed/src/main/scala/asobu/distributed/service/DRequestExtractor.scala:
```diff
@@ -0,0 +1,154 @@
+import scala.language.implicitConversions
+import scala.util.{Failure, Success, Try}
+
+object DRequestExtractor extends ExtractorFunctions {
```
One other thought/question, are the versions of cats/shapeless used here, going to clash potentially with what comes from core/common?

kailuowang committed:
minor remove conmment

nsauro commented:
ok..lets start down the path, see where we end up 👍
